doc/userguide/devguide/extending/app-layer/overview.rst

```diff
@@ -0,0 +1,102 @@
+**************************
+Application Layer Overview
+**************************
+
+.. contents:: Table of Contents
+
+This section aims to give an overview of what is needed to add
+an application-layer protocol to Suricata.
+
+After a generic first step of collecting data about this application-layer protocol,
+especially pcaps for testing, we can dive into the Suricata specifics.
+
+An application-layer protocol has three logic components in Suricata :
+- parser
+- logger
+- detection engine
+
+Both detection engine and logger will depend on the processing done by the parser.
+
+For security reasons, we now develop application-layer protocol code
+only in Rust and not in C.
+
+Parser
+======
+
+The parser is described by an instance of the structure ``RustParser``.
+
+A parser has :
+
+- a name (where it is better to avoid dashes)
+- an ipproto (if an app-layer is both over UDP and TCP, it needs to be registered with 2 RustParser)
+- flags : only one flag ``APP_LAYER_PARSER_OPT_ACCEPT_GAPS``
+- some app-layer detection logic see :ref:`Protocol detection`.
+- some logic around (one) State and (one) Transaction structures
+- some stringer functions (frames, events)
+
+So each app-layer protocol needs to define two structures : one State and one Transaction.
+A State will live throughout the flow (or until there is a protocol change in the flow).
+As such, it is useful to retain data that needs such a scope (for example HTTP2 dynamic headers table).
+And it is also useful if the parsing uses a state-machine logic, for example for file streaming.
+A State will own a list of :doc:`transactions`.
+
+Transactions
+------------
+
+Transactions are the basic logical unit used by Suricata for an application-layer protocol.
+
+The big decision is how to assemble PDUs into a transaction.
+Simplest design is to have one transaction per PDU (like DNS).
+But it may add value for example to combine request and response into a single transaction
+(like HTTP).
```

Comment:
Interesting. I always thought this was based on whether the protocol is stateless or stateful.

Reply:
Could you expand on this ?